// express-session-sqlite/build/SqliteStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SqliteStoreBase_1 = require("./SqliteStoreBase");
function sqliteStoreFactory(session) {
    const Store = session.Store;
    class SqliteStore extends Store {
        constructor(config) {
            super(config);
            this.sqliteStore = new SqliteStoreBase_1.SqliteStoreBase(config);
            this.cleanupTimer = setInterval(async () => {
                try {
                    await this.sqliteStore.removeExpiredSessions();
                }
                catch (e) {
                    // ignore
                }
                // 5 mins
            }, config.cleanupInterval || 300000);
        }
        get(sid, callback) {
            this.sqliteStore
                .get(sid)
                .then(data => {
                callback(null, data);
            })
                .catch(callback);
        }
        set(sid, session, callback) {
            this.sqliteStore
                .set(sid, session)
                .then(callback)
                .catch(callback);
        }
        destroy(sid, callback) {
            this.sqliteStore
                .destroy(sid)
                .then(callback)
                .catch(callback);
        }
        // @ts-ignore
        all(callback) {
            this.sqliteStore
                .all()
                .then(data => {
                callback(null, data);
            })
                .catch(callback);
        }
        length(callback) {
            this.sqliteStore
                .length()
                .then(length => {
                callback(null, length);
            })
                .catch(callback);
        }
        clear(callback) {
            this.sqliteStore
                .clear()
                .then(callback)
                .catch(callback);
        }
        touch(sid, session, callback) {
            this.sqliteStore
                .touch(sid, session)
                .then(callback)
                .catch(callback);
        }
    }
    // @ts-ignore
    return SqliteStore;
}
exports.default = sqliteStoreFactory;
